File: 240. Search a 2D Matrix II.c
/*
Write an efficient algorithm that searches for a value target in an m x n integer matrix matrix. This matrix has the following properties:

    Integers in each row are sorted in ascending from left to right.
    Integers in each column are sorted in ascending from top to bottom.

*/
bool binSearch(int* arr, int arrSize, int target){
    int l = 0;
    int r = arrSize-1;
    while(l < r){
        int m = l + (r-l)/2;
        if(arr[m] == target){
            return true;
        }else if(target < arr[m]){
            r = m-1;
        }else{
            l = m+1;
        }
    }
    return arr[l] == target;
}

bool _searchMatrix(int** matrix, int matrixSize, int* matrixColSize, int target, int y, int x){
    if(y == matrixSize-1){
        //printf("searching bottom row\n");
        return binSearch(matrix[y], matrixColSize[y], target);
        
    }else if(x == 0){
        //printf("searching left col\n");
        if(matrix[y][x] == target){
            return true;
        }
        return y < matrixSize && _searchMatrix(matrix, matrixSize, matrixColSize,target, y+1, x);
    }else if(matrix[y][x] == target){
        return true;
    }else if(target < matrix[y][x]){
        return _searchMatrix(matrix, matrixSize, matrixColSize,target, y, x-1);
    }else{
        return _searchMatrix(matrix, matrixSize, matrixColSize,target, y+1, x);
    }
}

bool searchMatrix(int** matrix, int matrixSize, int* matrixColSize, int target){
    int y = 0;
    int x = matrixColSize[matrixSize-1]-1;
    return _searchMatrix(matrix, matrixSize, matrixColSize,target, y, x);
}

File: 240. Search a 2D Matrix II.py
'''
Write an efficient algorithm that searches for a value target in an m x n integer matrix matrix. This matrix has the following properties:

    Integers in each row are sorted in ascending from left to right.
    Integers in each column are sorted in ascending from top to bottom.
'''
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        y,x = 0, len(matrix[0])-1
        
        def _searchMatrix(matrix, target, y, x):
            if y == len(matrix)-1:# search bottom row
                return target in matrix[y]
            elif x == 0:# search left col
                return target in [matrix[i][x] for i in range(y, len(matrix))]
            elif target == matrix[y][x]:
                return True
            elif target < matrix[y][x]:
                return _searchMatrix(matrix, target, y, x-1)
            else:
                return _searchMatrix(matrix, target, y+1, x)
            
        return _searchMatrix(matrix, target, y, x)
            
        
        